In [1]:
## This is the dataset of my friend alamgir
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
data = pd.read_csv('/content/ADNIMERGE_02May2025.csv')

In [2]:
data.head(10)

,RID,COLPROT,ORIGPROT,PTID,SITE,VISCODE,EXAMDATE,DX_bl,AGE,PTGENDER,...,PTAU_bl,FDG_bl,PIB_bl,AV45_bl,FBB_bl,Years_bl,Month_bl,Month,M,update_stamp
0,2,ADNI1,ADNI1,011_S_0002,11,bl,2005-09-08,CN,74.3,Male,...,NaN,1.33615,NaN,NaN,NaN,0.000000,0.00000,0.0,0.0,2023-07-07 04:59:40.0
1,3,ADNI1,ADNI1,011_S_0003,11,bl,2005-09-12,AD,81.3,Male,...,22.83,1.10860,NaN,NaN,NaN,0.000000,0.00000,0.0,0.0,2023-07-07 04:59:40.0
2,3,ADNI1,ADNI1,011_S_0003,11,m06,2006-03-13,AD,81.3,Male,...,22.83,1.10860,NaN,NaN,NaN,0.498289,5.96721,6.0,6.0,2023-07-07 04:59:40.0
3,3,ADNI1,ADNI1,011_S_0003,11,m12,2006-09-12,AD,81.3,Male,...,22.83,1.10860,NaN,NaN,NaN,0.999316,11.96720,12.0,12.0,2023-07-07 04:59:40.0
4,3,ADNI1,ADNI1,011_S_0003,11,m24,2007-09-12,AD,81.3,Male,...,22.83,1.10860,NaN,NaN,NaN,1.998630,23.93440,24.0,24.0,2023-07-07 04:59:40.0
5,4,ADNI1,ADNI1,022_S_0004,22,bl,2005-11-08,LMCI,67.5,Male,...,13.29,NaN,NaN,NaN,NaN,0.000000,0.00000,0.0,0.0,2023-07-07 04:59:40.0
6,4,ADNI1,ADNI1,022_S_0004,22,m06,2006-05-02,LMCI,67.5,Male,...,13.29,NaN,NaN,NaN,NaN,0.479124,5.73770,6.0,6.0,2023-07-07 04:59:40.0
7,4,ADNI1,ADNI1,022_S_0004,22,m12,2006-11-14,LMCI,67.5,Male,...,13.29,NaN,NaN,NaN,NaN,1.015740,12.16390,12.0,12.0,2023-07-07 04:59:40.0
8,4,ADNI1,ADNI1,022_S_0004,22,m18,2007-05-14,LMCI,67.5,Male,...,13.29,NaN,NaN,NaN,NaN,1.511290,18.09840,18.0,18.0,2023-07-07 04:59:40.0
9,4,ADNI1,ADNI1,022_S_0004,22,m36,2008-11-18,LMCI,67.5,Male,...,13.29,NaN,NaN,NaN,NaN,3.028060,36.26230,36.0,36.0,2023-07-07 04:59:40.0


In [3]:
# extracting the column for better understanding features
print(data.columns.tolist())

['RID', 'COLPROT', 'ORIGPROT', 'PTID', 'SITE', 'VISCODE', 'EXAMDATE', 'DX_bl', 'AGE', 'PTGENDER', 'PTEDUCAT', 'PTETHCAT', 'PTRACCAT', 'PTMARRY', 'APOE4', 'FDG', 'PIB', 'AV45', 'FBB', 'ABETA', 'TAU', 'PTAU', 'CDRSB', 'ADAS11', 'ADAS13', 'ADASQ4', 'MMSE', 'RAVLT_immediate', 'RAVLT_learning', 'RAVLT_forgetting', 'RAVLT_perc_forgetting', 'LDELTOTAL', 'DIGITSCOR', 'TRABSCOR', 'FAQ', 'MOCA', 'EcogPtMem', 'EcogPtLang', 'EcogPtVisspat', 'EcogPtPlan', 'EcogPtOrgan', 'EcogPtDivatt', 'EcogPtTotal', 'EcogSPMem', 'EcogSPLang', 'EcogSPVisspat', 'EcogSPPlan', 'EcogSPOrgan', 'EcogSPDivatt', 'EcogSPTotal', 'FLDSTRENG', 'FSVERSION', 'IMAGEUID', 'Ventricles', 'Hippocampus', 'WholeBrain', 'Entorhinal', 'Fusiform', 'MidTemp', 'ICV', 'DX', 'mPACCdigit', 'mPACCtrailsB', 'EXAMDATE_bl', 'CDRSB_bl', 'ADAS11_bl', 'ADAS13_bl', 'ADASQ4_bl', 'MMSE_bl', 'RAVLT_immediate_bl', 'RAVLT_learning_bl', 'RAVLT_forgetting_bl', 'RAVLT_perc_forgetting_bl', 'LDELTOTAL_BL', 'DIGITSCOR_bl', 'TRABSCOR_bl', 'FAQ_bl', 'mPACCdigit_bl

In [4]:
data['MMSE']

,MMSE
0,28.0
1,20.0
2,24.0
3,17.0
4,19.0
...,...
5219,30.0
5220,NaN
5221,NaN
5222,30.0


**Data preprocessing**

In [5]:
# Select baseline features (ending with '_bl') + PTID + DX (label)
baseline_cols = [col for col in data.columns if col.endswith('_bl')] + ['PTID', 'DX']
baseline_df = data[baseline_cols].copy()


In [6]:
baseline_df.head(10)
## We took only baseline features

,DX_bl,EXAMDATE_bl,CDRSB_bl,ADAS11_bl,ADAS13_bl,ADASQ4_bl,MMSE_bl,RAVLT_immediate_bl,RAVLT_learning_bl,RAVLT_forgetting_bl,...,TAU_bl,PTAU_bl,FDG_bl,PIB_bl,AV45_bl,FBB_bl,Years_bl,Month_bl,PTID,DX
0,CN,2005-09-08,0.0,10.67,18.67,5.0,28.0,44.0,4.0,6.0,...,NaN,NaN,1.33615,NaN,NaN,NaN,0.000000,0.00000,011_S_0002,CN
1,AD,2005-09-12,4.5,22.00,31.00,8.0,20.0,22.0,1.0,4.0,...,239.7,22.83,1.10860,NaN,NaN,NaN,0.000000,0.00000,011_S_0003,Dementia
2,AD,2005-09-12,4.5,22.00,31.00,8.0,20.0,22.0,1.0,4.0,...,239.7,22.83,1.10860,NaN,NaN,NaN,0.498289,5.96721,011_S_0003,Dementia
3,AD,2005-09-12,4.5,22.00,31.00,8.0,20.0,22.0,1.0,4.0,...,239.7,22.83,1.10860,NaN,NaN,NaN,0.999316,11.96720,011_S_0003,Dementia
4,AD,2005-09-12,4.5,22.00,31.00,8.0,20.0,22.0,1.0,4.0,...,239.7,22.83,1.10860,NaN,NaN,NaN,1.998630,23.93440,011_S_0003,Dementia
5,LMCI,2005-11-08,1.0,14.33,21.33,6.0,27.0,37.0,7.0,4.0,...,153.1,13.29,NaN,NaN,NaN,NaN,0.000000,0.00000,022_S_0004,MCI
6,LMCI,2005-11-08,1.0,14.33,21.33,6.0,27.0,37.0,7.0,4.0,...,153.1,13.29,NaN,NaN,NaN,NaN,0.479124,5.73770,022_S_0004,MCI
7,LMCI,2005-11-08,1.0,14.33,21.33,6.0,27.0,37.0,7.0,4.0,...,153.1,13.29,NaN,NaN,NaN,NaN,1.015740,12.16390,022_S_0004,MCI
8,LMCI,2005-11-08,1.0,14.33,21.33,6.0,27.0,37.0,7.0,4.0,...,153.1,13.29,NaN,NaN,NaN,NaN,1.511290,18.09840,022_S_0004,MCI
9,LMCI,2005-11-08,1.0,14.33,21.33,6.0,27.0,37.0,7.0,4.0,...,153.1,13.29,NaN,NaN,NaN,NaN,3.028060,36.26230,022_S_0004,MCI


#**Handling missing data**

In [7]:
# Handling the missing data
#  Checking missing ratio per column
missing_ratio = baseline_df.isnull().mean().sort_values(ascending=False)
print("Missing Value Ratio (Descending):")
print(missing_ratio)

# Dropping columns with more than 50% missing values
baseline_df = baseline_df.loc[:, missing_ratio <= 0.5]

print("\nAfter Dropping Columns with >50% Missing:")
print("Number of columns:", baseline_df.shape[1])

# Adding back important biomarkers manually if they were dropped
important_biomarkers = ['ABETA_bl', 'TAU_bl', 'PTAU_bl', 'FDG_bl', 'PIB_bl', 'AV45_bl', 'FBB_bl']

for col in important_biomarkers:
    if col not in baseline_df.columns and col in data.columns:
        baseline_df[col] = data[col]
        print(f"Added back important biomarker: {col}")

print("\nFinal Number of Columns after Cleaning:", baseline_df.shape[1])


Missing Value Ratio (Descending):
FBB_bl                      1.000000
PIB_bl                      0.989280
EcogSPOrgan_bl              0.562596
EcogPtOrgan_bl              0.562021
EcogSPVisspat_bl            0.553407
MOCA_bl                     0.553025
EcogSPDivatt_bl             0.553025
EcogSPPlan_bl               0.552259
EcogPtDivatt_bl             0.552067
EcogPtVisspat_bl            0.549005
AV45_bl                     0.548622
EcogPtPlan_bl               0.548239
EcogPtLang_bl               0.548047
EcogPtTotal_bl              0.548047
EcogSPMem_bl                0.547856
EcogPtMem_bl                0.547090
EcogSPTotal_bl              0.547090
EcogSPLang_bl               0.546899
DIGITSCOR_bl                0.456355
PTAU_bl                     0.340544
ABETA_bl                    0.340544
TAU_bl                      0.340544
FDG_bl                      0.276991
DX                          0.144525
Fusiform_bl                 0.115046
MidTemp_bl                  0.115046
Ento

<ipython-input-7-94441ac064f0>:18: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  baseline_df[col] = data[col]
<ipython-input-7-94441ac064f0>:18: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  baseline_df[col] = data[col]
<ipython-input-7-94441ac064f0>:18: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#

In [8]:
# Manually ensuring these biomarkers are kept
final_biomarkers = ['ABETA_bl', 'TAU_bl', 'PTAU_bl', 'FDG_bl']
baseline_df = baseline_df.loc[:, (baseline_df.isnull().mean() <= 0.5) | baseline_df.columns.isin(final_biomarkers)]

print("Final shape after keeping useful biomarkers:", baseline_df.shape)


Final shape after keeping useful biomarkers: (5224, 34)


**Missing value imputation**

In [9]:
# for Numerical features(Continuos) we will use KNN Imputer
# for Categorical features we will use mode imputation
from sklearn.impute import KNNImputer
from sklearn.preprocessing import LabelEncoder

# Separate numerical and categorical columns
num_cols = baseline_df.select_dtypes(include=['float64', 'int64']).columns
cat_cols = baseline_df.select_dtypes(include=['object', 'category']).columns

# Encode categorical variables temporarily (if any)
label_encoders = {}
for col in cat_cols:
    le = LabelEncoder()
    baseline_df[col] = baseline_df[col].astype(str)  # ensure it's string
    baseline_df[col] = le.fit_transform(baseline_df[col])
    label_encoders[col] = le

# Apply KNN Imputer for numerical data
imputer = KNNImputer(n_neighbors=5)
baseline_df[num_cols] = imputer.fit_transform(baseline_df[num_cols])



In [10]:
## Now checking the missing value
updated_missing_ratio = baseline_df.isnull().mean().sort_values(ascending=False)
print("Missing Value Ratio (Descending):")
print(updated_missing_ratio)

Missing Value Ratio (Descending):
DX_bl                       0.0
EXAMDATE_bl                 0.0
CDRSB_bl                    0.0
ADAS11_bl                   0.0
ADAS13_bl                   0.0
ADASQ4_bl                   0.0
MMSE_bl                     0.0
RAVLT_immediate_bl          0.0
RAVLT_learning_bl           0.0
RAVLT_forgetting_bl         0.0
RAVLT_perc_forgetting_bl    0.0
DIGITSCOR_bl                0.0
TRABSCOR_bl                 0.0
FAQ_bl                      0.0
mPACCdigit_bl               0.0
mPACCtrailsB_bl             0.0
FLDSTRENG_bl                0.0
FSVERSION_bl                0.0
IMAGEUID_bl                 0.0
Ventricles_bl               0.0
Hippocampus_bl              0.0
WholeBrain_bl               0.0
Entorhinal_bl               0.0
Fusiform_bl                 0.0
MidTemp_bl                  0.0
ICV_bl                      0.0
ABETA_bl                    0.0
TAU_bl                      0.0
PTAU_bl                     0.0
FDG_bl                      0.0
Years_

In [11]:
baseline_df.shape

(5224, 34)

**Removing Outlier**

In [12]:
# For outlier detection and removing , we use the IQR .
# we will keep the value which lies between 25 % quantile and 75 % quantile.

# store only numerical column

numerical_cols = baseline_df.select_dtypes(include=['float64', 'int64']).columns

# copying original dataset to keep a original dataset for further use

new_data = baseline_df.copy()

# Now we will calculate  the first quartile ( Q1) and third quartile (Q3) for each selected column
for col in numerical_cols:
    Q1 = baseline_df[col].quantile(0.25)
    Q3 = baseline_df[col].quantile(0.75)
    IQR = Q3 - Q1
    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR

    # keep data between IQR range
    new_data = new_data[(baseline_df[col] >= lower_bound) & (baseline_df[col] <= upper_bound)]


<ipython-input-12-d57869ddaa71>:21: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  new_data = new_data[(baseline_df[col] >= lower_bound) & (baseline_df[col] <= upper_bound)]
<ipython-input-12-d57869ddaa71>:21: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  new_data = new_data[(baseline_df[col] >= lower_bound) & (baseline_df[col] <= upper_bound)]
<ipython-input-12-d57869ddaa71>:21: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  new_data = new_data[(baseline_df[col] >= lower_bound) & (baseline_df[col] <= upper_bound)]
<ipython-input-12-d57869ddaa71>:21: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  new_data = new_data[(baseline_df[col] >= lower_bound) & (baseline_df[col] <= upper_bound)]
<ipython-input-12-d57869ddaa71>:21: UserWarning: Boolean Series key will be reindexed to match DataFrame index.
  new_data = new_data[(baseline_df[col] >= lower_bound) & (baseline_

In [13]:
print("this is the shape before outlier removal",baseline_df.shape)
print("this is the shape after outlier removal",new_data.shape)

this is the shape before outlier removal (5224, 34)
this is the shape after outlier removal (3408, 34)


# **Feature Selection**

In [14]:
## we will use lasso regression for our feature selection

## we will use lasso regression for our feature selection

# Define the target varible and features
x = new_data.drop(columns=['PTID', 'DX'])
y = new_data['DX']

In [15]:
## Now we will transform the data
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
x_scaled = scaler.fit_transform(x)

In [16]:
## Now we will use lasso for feature selection

from sklearn.linear_model import Lasso
from sklearn.model_selection import GridSearchCV
from sklearn.feature_selection import SelectFromModel

lasso = Lasso()
parameter = {'alpha': [0.001, 0.01, 0.1, 1, 10,20,50, 100]}
LassoCV = GridSearchCV(lasso, parameter, scoring='neg_mean_squared_error', cv=5)
LassoCV.fit(x_scaled, y)



GridSearchCV(cv=5, estimator=Lasso(),
             param_grid={'alpha': [0.001, 0.01, 0.1, 1, 10, 20, 50, 100]},
             scoring='neg_mean_squared_error')

In [17]:
print('Best alpha:', LassoCV.best_params_)
print('Best parameter',LassoCV.best_score_)

Best alpha: {'alpha': 0.001}
Best parameter -0.6824767447287392


In [18]:
# we get the best alpha value is 0.1. Now we will again fit lasso with this alpha value
new_lasso = Lasso(alpha=0.1)
new_lasso.fit(x_scaled, y)

# Select best features
lasso_features = x.columns[new_lasso.coef_ != 0]
print("Best features:", lasso_features)
print(lasso_features.to_list())

Best features: Index(['DX_bl', 'CDRSB_bl', 'mPACCdigit_bl', 'FLDSTRENG_bl', 'FSVERSION_bl'], dtype='object')
['DX_bl', 'CDRSB_bl', 'mPACCdigit_bl', 'FLDSTRENG_bl', 'FSVERSION_bl']


In [19]:
## we will use Random Forest feature Importance for cross checking

from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import cross_val_score

rf = RandomForestClassifier()
rf.fit(x_scaled, y)

importance = pd.Series(rf.feature_importances_, index=x.columns)
rf_features = importance[importance>importance.mean()].index
print("Best features:", rf_features)
print(rf_features.to_list())


Best features: Index(['DX_bl', 'CDRSB_bl', 'FAQ_bl', 'mPACCdigit_bl', 'mPACCtrailsB_bl',
       'Years_bl', 'Month_bl'],
      dtype='object')
['DX_bl', 'CDRSB_bl', 'FAQ_bl', 'mPACCdigit_bl', 'mPACCtrailsB_bl', 'Years_bl', 'Month_bl']


In [20]:
## Now we will Combine lasso_features and rf_features

final_features = set(lasso_features) | set(rf_features)
print("Final features:")
print(final_features)


Final features:
{'FLDSTRENG_bl', 'Years_bl', 'CDRSB_bl', 'FSVERSION_bl', 'DX_bl', 'FAQ_bl', 'Month_bl', 'mPACCdigit_bl', 'mPACCtrailsB_bl'}


In [21]:
# from domain knowledge we should those variable also 'AGE', 'APOE4', 'CDRSB', 'FAQ', 'FDG', 'MMSE', 'PTEDUCAT','RAVLT_forgetting', 'RAVLT_immediate', 'RAVLT_learning', 'RAVLT_percent_forgetting', 'TRABSCOR', 'blTAU', 'blTRABSCOR'

#Now we merge those features with our previous feature and we get --->

# Check available columns in new_data
available_columns = new_data.columns.tolist()

# Select features that are present in new_data
selected_features = [col for col in ['AGE', 'APOE4', 'CDRSB', 'FAQ', 'FDG', 'MMSE', 'PTEDUCAT',
                     'RAVLT_forgetting', 'RAVLT_immediate', 'RAVLT_learning',
                     'RAVLT_percent_forgetting', 'TRABSCOR', 'blTAU', 'blTRABSCOR','mPACCdigit_bl', 'Month_bl', 'CDRSB_bl','Years_bl', 'mPACCtrailsB_bl'] if col in available_columns]


## new define x and y

x = new_data[selected_features]
y = new_data['DX']

# **Prediction**

In [22]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

## Train test split

xtrain, xtest, ytrain, ytest = train_test_split(x, y, test_size=0.2, random_state=51)

## features scaling

scaler = StandardScaler()
xtrain = scaler.fit_transform(xtrain)
xtest = scaler.transform(xtest)


**Hyper Parameter Tuning with Optuna**

In [23]:
!pip install catboost
# Importing the required libraries
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from sklearn.svm import SVC
from xgboost import XGBClassifier
from lightgbm import LGBMClassifier
from catboost import CatBoostClassifier


   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 99.2/99.2 MB 8.5 MB/s eta 0:00:00


In [24]:
## Using optuna for hyper parameter tuning.
## here we use GradientBoosting,RandomForest,XGBoost,CatBoost for model

def objective(trial):
    classifier_name = trial.suggest_categorical('classifier', ['GradientBoosting','RandomForest', 'XGBoost','CatBoost'])

    if classifier_name == 'GradientBoosting':
        n_estimators = trial.suggest_int('n_estimators', 50, 300)
        learning_rate = trial.suggest_float('learning_rate', 0.01, 0.3, log=True)
        max_depth = trial.suggest_int('max_depth', 3, 20)
        min_samples_split = trial.suggest_int('min_samples_split', 2, 10)
        min_samples_leaf = trial.suggest_int('min_samples_leaf', 1, 10)

        model = GradientBoostingClassifier(
            n_estimators=n_estimators,
            learning_rate=learning_rate,
            max_depth=max_depth,
            min_samples_split=min_samples_split,
            min_samples_leaf=min_samples_leaf,
            random_state=42
        )
    elif classifier_name == 'RandomForest': # Corrected indentation here
        # Random Forest hyperparameters
        n_estimators = trial.suggest_int('n_estimators', 50, 300)
        max_depth = trial.suggest_int('max_depth', 3, 20)
        min_samples_split = trial.suggest_int('min_samples_split', 2, 10)
        min_samples_leaf = trial.suggest_int('min_samples_leaf', 1, 10)
        bootstrap = trial.suggest_categorical('bootstrap', [True, False])

        model = RandomForestClassifier(
            n_estimators=n_estimators,
            max_depth=max_depth,
            min_samples_split=min_samples_split,
            min_samples_leaf=min_samples_leaf,
            bootstrap=bootstrap,
            random_state=42
        )

    elif classifier_name == 'XGBoost':
        n_estimators = trial.suggest_int('n_estimators', 50, 300)
        max_depth = trial.suggest_int('max_depth', 3, 15)
        learning_rate = trial.suggest_float('learning_rate', 0.01, 0.3, log=True)
        subsample = trial.suggest_float('subsample', 0.5, 1.0)
        colsample_bytree = trial.suggest_float('colsample_bytree', 0.5, 1.0)

        model = XGBClassifier(
            n_estimators=n_estimators,
            max_depth=max_depth,
            learning_rate=learning_rate,
            subsample=subsample,
            colsample_bytree=colsample_bytree,
            use_label_encoder=False,
            eval_metric='mlogloss',
            random_state=42
        )

    elif classifier_name == 'CatBoost':
        depth = trial.suggest_int("depth", 3, 10)
        learning_rate = trial.suggest_float("learning_rate", 0.01, 0.3, log=True)
        iterations = trial.suggest_int("iterations", 100, 500)

        model = CatBoostClassifier(
            depth=depth,
            learning_rate=learning_rate,
            iterations=iterations,
            verbose=0,
            random_state=42
        )

    score = cross_val_score(model, xtrain, ytrain, cv=5, scoring='accuracy').mean()
    return score

In [25]:
!pip install optuna
import optuna

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 386.6/386.6 kB 10.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 231.9/231.9 kB 16.8 MB/s eta 0:00:00


In [26]:
# Create a study and optimize it
study = optuna.create_study(direction='maximize')
study.optimize(objective, n_trials=50)

[I 2025-05-16 00:45:57,963] A new study created in memory with name: no-name-29cae498-91a8-4949-a89e-c8ee0e3986af
[I 2025-05-16 00:46:52,710] Trial 0 finished with value: 0.8576751688678292 and parameters: {'classifier': 'GradientBoosting', 'n_estimators': 78, 'learning_rate': 0.0365635494486751, 'max_depth': 15, 'min_samples_split': 3, 'min_samples_leaf': 5}. Best is trial 0 with value: 0.8576751688678292.
[I 2025-05-16 00:46:54,288] Trial 1 finished with value: 0.8209900191551569 and parameters: {'classifier': 'RandomForest', 'n_estimators': 87, 'max_depth': 14, 'min_samples_split': 4, 'min_samples_leaf': 9, 'bootstrap': True}. Best is trial 0 with value: 0.8576751688678292.
[I 2025-05-16 00:47:51,739] Trial 2 finished with value: 0.8664751150989684 and parameters: {'classifier': 'GradientBoosting', 'n_estimators': 287, 'learning_rate': 0.015687834861734033, 'max_depth': 7, 'min_samples_split': 6, 'min_samples_leaf': 6}. Best is trial 2 with value: 0.8664751150989684.
[I 2025-05-16 0

In [27]:
# Retrieve the best trial
best_trial = study.best_trial
print("Best trial parameters:", best_trial.params)
print("Best trial accuracy:", best_trial.value)

Best trial parameters: {'classifier': 'GradientBoosting', 'n_estimators': 154, 'learning_rate': 0.04367705372907881, 'max_depth': 8, 'min_samples_split': 4, 'min_samples_leaf': 6}
Best trial accuracy: 0.869043250327654


In [28]:
## Find the best hyperparameters for each model
import pandas as pd

#Converting Optuna trials to DataFrame
df_trials = study.trials_dataframe()

# Filtering best trials for each model
gb = df_trials[df_trials['params_classifier'] == 'GradientBoosting']
rf = df_trials[df_trials['params_classifier'] == 'RandomForest']
xgb = df_trials[df_trials['params_classifier'] == 'XGBoost']
cb = df_trials[df_trials['params_classifier'] == 'CatBoost']

# Get best performing trial for each model
best_gb = gb.sort_values('value', ascending=False).iloc[0]
best_rf = rf.sort_values('value', ascending=False).iloc[0]
best_xgb = xgb.sort_values('value', ascending=False).iloc[0]
best_cb = cb.sort_values('value', ascending=False).iloc[0]

#Extract only hyperparameters from each trial
gb_params = {k.replace('params_', ''): v for k, v in best_gb.items() if k.startswith('params_')}
rf_params = {k.replace('params_', ''): v for k, v in best_rf.items() if k.startswith('params_')}
xgb_params = {k.replace('params_', ''): v for k, v in best_xgb.items() if k.startswith('params_')}
cb_params = {k.replace('params_', ''): v for k, v in best_cb.items() if k.startswith('params_')}

#Print best hyperparameters
print(" Best GradientBoosting Parameters:\n", gb_params)
print("\n Best RandomForest Parameters:\n", rf_params)
print("\n Best XGBoost Parameters:\n", xgb_params)
print("\n Best CatBoost Parameters:\n", cb_params)


 Best GradientBoosting Parameters:
 {'bootstrap': nan, 'classifier': 'GradientBoosting', 'colsample_bytree': np.float64(nan), 'depth': np.float64(nan), 'iterations': np.float64(nan), 'learning_rate': np.float64(0.04367705372907881), 'max_depth': np.float64(8.0), 'min_samples_leaf': np.float64(6.0), 'min_samples_split': np.float64(4.0), 'n_estimators': np.float64(154.0), 'subsample': np.float64(nan)}

 Best RandomForest Parameters:
 {'bootstrap': True, 'classifier': 'RandomForest', 'colsample_bytree': np.float64(nan), 'depth': np.float64(nan), 'iterations': np.float64(nan), 'learning_rate': np.float64(nan), 'max_depth': np.float64(15.0), 'min_samples_leaf': np.float64(4.0), 'min_samples_split': np.float64(8.0), 'n_estimators': np.float64(138.0), 'subsample': np.float64(nan)}

 Best XGBoost Parameters:
 {'bootstrap': nan, 'classifier': 'XGBoost', 'colsample_bytree': np.float64(0.5418688503815253), 'depth': np.float64(nan), 'iterations': np.float64(nan), 'learning_rate': np.float64(0.1878

# **Stacking**

In [29]:
## Now we will use stacking for improve the accuracy with the help of our model with those best hyper parameter

from sklearn.linear_model import LogisticRegression
## tuning with their best hyper parameter
from sklearn.ensemble import StackingClassifier
# Gradient Boosting
gb_model = GradientBoostingClassifier(
    n_estimators=105,
    learning_rate=0.04367705372907881,
    max_depth=8,
    min_samples_split=4,
    min_samples_leaf=6,
    random_state=51
)

# Random Forest
rf_model = RandomForestClassifier(
    n_estimators=138,
    max_depth=15,
    min_samples_split=8,
    min_samples_leaf=4,
    bootstrap=True,
    random_state=51
)

# XGBoost
xgb_model = XGBClassifier(
    n_estimators=80,
    max_depth=5,
    learning_rate=0.1878293909738388,
    subsample=0.9902078958016296,
    colsample_bytree=0.5418688503815253,
    use_label_encoder=False,
    eval_metric='mlogloss',
    random_state=51
)

# CatBoost
cb_model = CatBoostClassifier(
    iterations=462,
    depth=10,
    learning_rate=0.126551199514587,
    verbose=0,
    random_state=51
)


In [30]:
estimators = [
    ('gb', gb_model),
    ('rf', rf_model),
    ('xgb', xgb_model),
    ('cb', cb_model)
]

stacking_model = StackingClassifier(
    estimators=estimators,
    final_estimator=LogisticRegression(),
    cv=5
)


In [31]:
stacking_model.fit(xtrain, ytrain)

from sklearn.metrics import accuracy_score

y_pred = stacking_model.predict(xtest)
acc = accuracy_score(ytest, y_pred)
print(f" Stacking Classifier Accuracy: {acc:.4f}")


/usr/local/lib/python3.11/dist-packages/xgboost/core.py:158: UserWarning: [01:29:49] WARNING: /workspace/src/learner.cc:740: 
Parameters: { "use_label_encoder" } are not used.

  warnings.warn(smsg, UserWarning)
/usr/local/lib/python3.11/dist-packages/xgboost/core.py:158: UserWarning: [01:30:44] WARNING: /workspace/src/learner.cc:740: 
Parameters: { "use_label_encoder" } are not used.

  warnings.warn(smsg, UserWarning)


 Stacking Classifier Accuracy: 0.8666


In [32]:
## Now we are using stacking with the same model and set meta model as gradientboosting model

from sklearn.ensemble import StackingClassifier, GradientBoostingClassifier

# Final estimator (meta-learner)
final_estimator = GradientBoostingClassifier(
    n_estimators=100,
    learning_rate=0.1,
    max_depth=3,
    random_state=42
)

# Define stacking model
stacking_model = StackingClassifier(
    estimators=estimators,
    final_estimator=final_estimator,
    cv=5
)



In [33]:
stacking_model.fit(xtrain, ytrain)

y_pred = stacking_model.predict(xtest)
acc = accuracy_score(ytest, y_pred)
print(f" Stacking Classifier Accuracy: {acc:.4f}")

/usr/local/lib/python3.11/dist-packages/xgboost/core.py:158: UserWarning: [01:34:09] WARNING: /workspace/src/learner.cc:740: 
Parameters: { "use_label_encoder" } are not used.

  warnings.warn(smsg, UserWarning)
/usr/local/lib/python3.11/dist-packages/xgboost/core.py:158: UserWarning: [01:35:08] WARNING: /workspace/src/learner.cc:740: 
Parameters: { "use_label_encoder" } are not used.

  warnings.warn(smsg, UserWarning)
/usr/local/lib/python3.11/dist-packages/xgboost/core.py:158: UserWarning: [01:35:09] WARNING: /workspace/src/learner.cc:740: 
Parameters: { "use_label_encoder" } are not used.

  warnings.warn(smsg, UserWarning)


 Stacking Classifier Accuracy: 0.8680


# **Conclusion**
So we find the best algorithm which is **GradientBoosting** with hyperparameter  (
    n_estimators=105,
    learning_rate=0.04367705372907881,
    max_depth=8,
    min_samples_split=4,
    min_samples_leaf=6
)

The accuracy was **0.869043250327654** .